src/content/developers/docs/smart-contracts/formal-verification/index.md

```diff
@@ -0,0 +1,134 @@
+
+Specifications are a collection of properties used to describe the intended behavior of a smart contract. Formal verification can prove that a contract behaves exactly as described in the specification. Unlike testing based on informal assumptions, formal verification provides mathematical proof of a smart contract's correctness.
+
+Formal verification is critical for smart contracts because minor vulnerabilities can lead to massive losses. For example, the [Parity Multisig wallet hack](https://www.freecodecamp.org/news/a-hacker-stole-31m-of-ether-how-it-happened-and-what-it-means-for-ethereum-9e5dc29e33ce/amp/) was the result of an unintended function that allowed the hacker to override owner settings and drain the contract of up to $31 million in ETH. 
+
+## How does formal verification work? {#how-does-formal-verification-work}
+
+Formal verification of Ethereum smart contracts can either be executed at the language level or bytecode level or both. Here’s a quick overview of both approaches:
+
+### High-level verification {#high-level-verification}
+
+Source code for Ethereum smart contracts is usually [written in high-level programming languages](/developers/docs/smart-contracts/languages/), such as Solidity, Vyper, or Yul. The source code defines the business logic, which determines how users can interact with the contract and how the contract should behave in response to certain inputs.
+
+Formal verification can help verify the correctness of a contract's source code. This requires transforming the business logic into a set of formalized, high-level specifications. These specifications must adequately and comprehensively express the desired functionality of the smart contract.
+
+The actual verification at this stage consists of checking the source code and confirming that it satisfies all specifications and requirements. Therefore, high-level verification gives developers a high degree of confidence that the source code has no bugs.
+
+### Low-level verification {#low-level-verification}
+
+Smart contracts written in high-level languages are  compiled into **bytecode**—low-level machine instructions that the Ethereum Virtual Machine (EVM) understands. Source code may undergo subtle changes during compilation, which may cause the contract to execute operations not defined in high-level specifications.
+
+Formal verification can help determine the correctness of a contract's compiled bytecode. This requires creating bytecode-based, low-level specifications that define how a smart contract should behave in the EVM environment.
+
+Low-level formal verification uses techniques like static analysis and fuzzing to ensure that contract bytecode satisfies all properties. This assures developers that the business logic specified in the source code is what will run in the context of the EVM.
+
+## Advantages and disadvantages of formal verification {#advantages-and-disadvantages-of-formal-verification}
```

Comment:
Would break this header up... felt a little awkward reading and having the numbered list go 1. 2. 3. 1. 2. 3

```suggestion
## Advantages of formal verification {#advantages-of-formal-verification}
```